syntax = "proto3";

package bilibili.broadcast.v1;

import "bilibili/rpc/status.proto";
import "google/protobuf/any.proto";
import "github.com/gogo/protobuf/gogoproto/gogo.proto";

option go_package = "git.bilibili.co/bapis/bapis-go/bilibili/broadcast/v1;v1";
option java_multiple_files = true;
option java_package = "com.bapis.bilibili.broadcast.v1";
option objc_class_prefix = "BAPIBroadcastV1";

option (gogoproto.sizer_all) = false;
option (gogoproto.marshaler_all) = false;
option (gogoproto.unmarshaler_all) = false;

message RoomJoinEvent {}
message RoomLeaveEvent {}
message RoomOnlineEvent {
    int32 online = 1;
    int32 all_online = 2;
}
message RoomMessageEvent {
    string target_path = 1;
    google.protobuf.Any body = 2;
}
message RoomErrorEvent {
    bilibili.rpc.Status status = 1;
}

message RoomReq {
    // {type}://{room_id}
    string id = 1;
    oneof event {
        RoomJoinEvent join = 2;
        RoomLeaveEvent leave = 3;
        RoomOnlineEvent online = 4;
        RoomMessageEvent msg = 5;
    }
}
message RoomResp {
    // {type}://{room_id}
    string id = 1;
    oneof event {
        RoomJoinEvent join = 2;
        RoomLeaveEvent leave = 3;
        RoomOnlineEvent online = 4;
        RoomMessageEvent msg = 5;
        RoomErrorEvent err = 6;
    }
}

service BroadcastRoom {
    rpc Enter(stream RoomReq) returns (stream RoomResp);
}
